File: python/code/79.py
class Solution:
    def dfs(self,i, j, board, word):
        if len(word) == 0:
            return True
        if i < 0 or j < 0 or i >= self.n or j >= self.m or board[i][j] != word[0] or self.visited[i][j]:
            return False
        if board[i][j] == word[0]:
            self.visited[i][j] = True
            word_ = word[1:]
            if self.dfs(i+1, j, board, word_) or self.dfs(i, j+1, board, word_) or self.dfs(i-1, j, board, word_) or self.dfs(i, j-1, board, word_):
                return True
            self.visited[i][j] = False
        return False

    def exist(self, board, word):
        self.n = len(board)
        if self.n == 0:
            return False
        self.m = len(board[0])
        self.visited = [[False for i in range(self.m)] for j in range(self.n)]
        for i in range(self.n):
            for j in range(self.m):
                if board[i][j] == word[0]:
                    if self.dfs(i, j, board, word):
                        return True
        return False

File: C++/code/79.cpp
class Solution {
public:
    void dfs(int x, int y, string word) {
        if (found || x < 0 || x >= m || y < 0 || y >= n || m_board[x][y] != word[0]) return;
        m_board[x][y] = '#';
        if (word.length() == 1) {
            found = true;
            return;
        }
        for (auto& dire: dires) {
            dfs(x + dire[0], y + dire[1], word.substr(1));
        }
        m_board[x][y] = word[0];
    }
    
    bool exist(vector<vector<char>>& board, string word) {
        m_board = board;
        m = board.size();
        n = board[0].size();
        found = false;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                dfs(i, j, word);
            }
        }
        return found;
    }
    
    vector<vector<int>> dires{{1,0},{-1,0},{0,1},{0,-1}};
    bool found;
    vector<vector<char>> m_board;
    int m, n;
};